openedx_tagging/core/tagging/models/system_defined.py

```diff
@@ -155,41 +58,101 @@ class ModelSystemDefinedTaxonomy(SystemDefinedTaxonomy):
+            self.tag_class_model.objects.get(**{f"{self.tag_class_value_field}__iexact": value})
+            return True
+        except ObjectDoesNotExist:
+            return False
+
+    def tag_for_value(self, value: str):
+        """
+        Get the Tag object for the given value.
+        """
+        try:
+            # First we look up the instance by value.
+            # We specify 'iexact' but whether it's case sensitive or not on MySQL depends on the model's collation.
+            instance = self.tag_class_model.objects.get(**{f"{self.tag_class_value_field}__iexact": value})
+        except ObjectDoesNotExist:
+            raise Tag.DoesNotExist
+        # Use the canonical value from here on (possibly with different case from the value given as a parameter)
+        value = getattr(instance, self.tag_class_value_field)
+        # We assume the value may change but the external_id is immutable.
+        # So look up keys using external_id. There may be a key with the same external_id but an out of date value.
+        external_id = str(getattr(instance, self.tag_class_key_field))
+        tag, _created = self.tag_set.get_or_create(external_id=external_id, defaults={"value": value})
+        if tag.value != value:
+            # Update the Tag to reflect the new cached 'value'
+            tag.value = value
+            tag.save()
+        return tag
+
+    def validate_external_id(self, external_id: str):
+        """
+        Check if 'external_id' is part of this Taxonomy.
+        """
+        try:
+            self.tag_class_model.objects.get(**{f"{self.tag_class_key_field}__iexact": external_id})
+            return True
+        except ObjectDoesNotExist:
+            return False
+
+    def tag_for_external_id(self, external_id: str):
+        """
+        Get the Tag object for the given external_id.
+        Some Taxonomies may auto-create the Tag at this point, e.g. a User
+        Taxonomy will create User Tags "just in time".
+
+        Will raise Tag.DoesNotExist if the tag is not valid for this taxonomy.
+        """
+        try:
+            # First we look up the instance by external_id
+            # We specify 'iexact' but whether it's case sensitive or not on MySQL depends on the model's collation.
+            instance = self.tag_class_model.objects.get(**{f"{self.tag_class_key_field}__iexact": external_id})
+        except ObjectDoesNotExist:
+            raise Tag.DoesNotExist
+        value = getattr(instance, self.tag_class_value_field)
+        # Use the canonical external_id from here on (may differ in capitalization)
+        external_id = getattr(instance, self.tag_class_key_field)
+        tag, _created = self.tag_set.get_or_create(external_id=external_id, defaults={"value": value})
+        if tag.value != value:
+            # Update the Tag to reflect the new cached 'value'
+            tag.value = value
+            tag.save()
+        return tag
 
 
-class UserModelObjectTag(ModelObjectTag):
```

Comment:
Since we're not subclassing ObjectTag for system-defined taxonomies anymore (yay!), the [System Taxonomy creation ADR](https://github.com/openedx/openedx-learning/blob/braden/simplify-tag-models/docs/decisions/0012-system-taxonomy-creation.rst#system-tag-lists-and-validation) should be updated to describe the use of Taxonomy subclasses instead.

Reply:
Good point. I updated it. I'm going to merge now but feel free to leave comments on this diff if you have any, and I'll address them in future PRs.